base/asyncmap.jl

```diff
@@ -62,24 +72,34 @@ function done(itr::AsyncCollector, state::AsyncCollectorState)
 end
 
 function next(itr::AsyncCollector, state::AsyncCollectorState)
-    # Wait if the maximum number of concurrent tasks are already running
+    # Wait if the maximum number of concurrent tasks are already running.
     while isbusy(itr, state)
         wait(state)
     end
 
-    # Get index and mapped function arguments from enumeration iterator
+    # Get index and mapped function arguments from enumeration iterator.
     (i, args), state.enum_state = next(itr.enumerator, state.enum_state)
 
-    # Execute function call and save result asynchronously
+    # Execute function call and save result asynchronously.
     @async begin
-        itr.results[i] = itr.f(args...)
-        state.active_count -= 1
+        try
+            itr.results[i] = itr.f(args...)
+        catch e
+            state.in_error = true
+            notify(state.task_done, e; error=true)
```

Comment:
That `notify` stops the processing immediately, even if it means avoiding at most one extra item being processed. To make the code clearer we could have a single `notify` as you suggest as `notify(state.task_done, state.in_error)` and 
and change the `wait` statement to `wait(state) && return (nothing, state)` with an appropriate comment.